contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,245 @@
+use std::fmt::{self, Display, Formatter};
+use std::marker::{Send, Sized};
+
+use rocket::config::{self, Value};
+
+pub use rocket_contrib_codegen::database;
+
+use self::r2d2::ManageConnection;
+
+#[cfg(any(feature = "diesel_sqlite_pool", feature = "diesel_postgres_pool", feature = "diesel_mysql_pool"))]
+pub extern crate diesel;
+
+#[cfg(feature = "postgres_pool")]
+pub extern crate postgres;
+#[cfg(feature = "postgres_pool")]
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "mysql_pool")]
+pub extern crate mysql;
+#[cfg(feature = "mysql_pool")]
+pub extern crate r2d2_mysql;
+
+#[cfg(feature = "sqlite_pool")]
+pub extern crate rusqlite;
+#[cfg(feature = "sqlite_pool")]
+pub extern crate r2d2_sqlite;
+
+#[cfg(feature = "cypher_pool")]
+pub extern crate rusted_cypher;
+#[cfg(feature = "cypher_pool")]
+pub extern crate r2d2_cypher;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+///
+/// Example configuration:
+/// ```toml
+/// [[global.databases]]
+/// name = "my_database"
+/// url = "postgres://postgres@localhost/my_database
+/// pool_size = 4
+/// ```
+pub struct DatabaseConfig<'a> {
+    /// The connection URL specified in the Rocket configuration
+    pub url: &'a str,
+    /// The size of the pool to be initialized. Defaults to the number of Rocket workers.
+    pub pool_size: i64,
+    /// Any extra options that are included in the configuration.
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+#[derive(Debug)]
+pub enum DatabaseConfigError {
+    MissingConfig,
+    MissingTable(String),
+    MissingConnectionUrl,
+    InvalidPoolSize(i64),
+}
+
+impl DatabaseConfigError {
+    pub fn pretty_print<'a>(&'a self, database_name: &'a str) {
+        error!("Error while instantiating database: '{}': {}", database_name, self);
+    }
+}
+
+impl<'a> Display for DatabaseConfigError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            DatabaseConfigError::MissingConfig => {
+                write!(f, "The active configuration does not contain a `databases` table.")
+            },
+            DatabaseConfigError::MissingTable(invalid_name) => {
+                write!(f, "Database-specific configuration was not found that was named: '{}'", invalid_name)
+            },
+            DatabaseConfigError::MissingConnectionUrl => {
+                write!(f, "The specified connection url is invalid")
```

Comment:
Is it invalid or is it not present?

Reply:
So, that's the difficult part. I think it could be both.

It could be missing, which would spawn this error, but it could also not be a valid string. For example, I just tested it and the error shows up when the `url` is missing and when it's set to a ridiculous value like `17`.

So, perhaps a clearer error message includes the possibility that it's missing as well? I'm not sure there's a straightforward way in the current error structure to clearly distinguish between invalid and not present. Conversely, we could add a new error type and spawn the error earlier.